### DISCUSSION TITLE
The App Dev Perspective: Howto write a container app that uses bpfman on a stand-alone OCI container host from containers (not k8s)?

### DISCUSSION THREAD

**Discussion by thediveo** [bpfman/bpfman#1069] [Q&A]
I would like to explore how a bpfman infrastructure feels on a "stand-alone" OCI container host with containerized applications using bpfman services, but am stuck in lots of voids that need more filling in:
- what needs to be integrated into the host itself? Presumably the bpfman-rpc service, socket-activated by a system manager such as systemd?
- coming from a strong Go background (including cilium/ebpf), what does a containerized application actually need to import and call in order to set up? Kind of a developer's welcome guide and walkthrough.

Now, I've seen and read at least a little bit through the examples, in addition to the official web site. What I'm noticing here is while there is a "dev" section, this addresses tinkering with the pdfman code base. Also, there's quite some information regarding k8s environments, but sparse information about starting small with stand-alone container hosts without k8s.

Would you be interested in helping me and getting feedback on such a "container app developer journey"? Would you like to eventually bring such information to your project's web site? And also discuss aspects of the current state of the code base, because I happen to have kind of a deja-vue here from the podman code base, regarding potential problems that ordinary app developers face (especially difficult codebase imports)?

**Comment by astoycos** [comment]
Hiya @thediveo!! 

> Would you be interested in helping me and getting feedback on such a "container app developer journey"? Would you like to eventually bring such information to your project's web site? And also discuss aspects of the current state of the code base, because I happen to have kind of a deja-vue here from the podman code base, regarding potential problems that ordinary app developers face (especially difficult codebase imports)?

We'd 100% be interested in fleshing out the non k8s container app developer journey and have some ideas on how to make this better (bpf token management feature etc).  We've basically focused on the simpler non-containerized process on the host OR the K8s use cases so there's some gaps with the "container on host" user story. 

> what needs to be integrated into the host itself? Presumably the bpfman-rpc service, socket-activated by a system manager such as systemd?

Yep so today your container application would use the rpc api exposed by `bpfman-rpc` that binary should be managed by systemd so that we get the benefits of socket activation (i.e a smaller dangerous permission attach surface) however it could also be started directly on the host with `sudo ./bpfman-rpc --timeout=0` for quicker development. The socket for communication with the rpc server via unix would also need to be shared with the container. Lastly, the eBPF map pins will need to be shared with the container via host mounts (in k8s this is done via CSI(the Container Storage Interface)), it'd be great if we could find some better mechanism for map sharing for the "stand-alone" OCI container host story.

> coming from a strong Go background (including cilium/ebpf), what does a containerized application actually need to import and call in order to set up? Kind of a developer's welcome guide and walkthrough.

Most of us are :) so all the application would need to import would be the  [generated go-grpc bindings](https://github.com/bpfman/bpfman/tree/main/clients/gobpfman/v1) for our grpc API.  It'd be great if we had simpler developer docs for this that aren't directly concerning the k8s use case.
